### PR TITLE
Made it easier to enter edit mode and to bring up the menu

### PR DESCRIPTION
**Pull request by michaelcvoigt**
Just say 'Edit .... '
Just say 'Debug menu ... '

### CODE REVIEW

File: Assets/Editor/Tools/EditorWindow/Voice/VoiceCommandEditorWindow.cs [hunk 1]
```diff
@@ -85,6 +85,26 @@ private void DrawCommands()
                 }
                 EditorGUILayout.EndHorizontal();
             }
+
+            /// <summary>
+            /// show buttons for the unprotected commands regesitered via DebugService.cs
+            /// </summary>
```
Comment (thegoldenmule) [inline Assets/Editor/Tools/EditorWindow/Voice/VoiceCommandEditorWindow.cs]: The XMLDoc comment style should only be used on methods and classes. Inside of methods and classes, just use `//`.

File: Assets/Editor/Tools/EditorWindow/Voice/VoiceCommandEditorWindow.cs [hunk 2]
```diff
@@ -85,6 +85,26 @@ private void DrawCommands()
...
+           
+            var unprotectedKeys = manager.UnprotectedCallbacks.Keys.ToArray();
+
```
Comment (thegoldenmule) [inline Assets/Editor/Tools/EditorWindow/Voice/VoiceCommandEditorWindow.cs]: Lots of random newlines throughout this PR. These should be removed throughout.

File: Assets/Editor/Tools/EditorWindow/Voice/VoiceCommandEditorWindow.cs (hunk 1 above)
Comment (thegoldenmule) [reply Assets/Editor/Tools/EditorWindow/Voice/VoiceCommandEditorWindow.cs]: FYI there are a few instances of this, but I'll just comment on this one.

File: Assets/Source/Application/States/Play/Hmd/Controllers/HmdDesignController.cs [hunk 3]
```diff
@@ -405,9 +405,12 @@ private void SetupPlay()
 
             _primaryAnchor.Setup();
 
-            _voice.RegisterAdmin("menu", Voice_OnPlayMenu);
-            _voice.RegisterAdmin("edit", Voice_OnEdit);
+            // Made this easier to say - Just say 'Debug menu...;
+            _voice.Register("menu", Voice_OnPlayMenu);
 
+            // Added unprotected way to enter edit mode - Just say 'Edit...'
+            _voice.RegisterUnprotected("edit", Voice_OnEdit);
```
Comment (thegoldenmule) [inline Assets/Source/Application/States/Play/Hmd/Controllers/HmdDesignController.cs]: This will have already been added in the `DebugService`. I think we should probably remove the `DebugService` one, as the `edit` voice command is not always valid (like if we are already in edit mode, or in some non-play mode).

File: Assets/Source/Input/Voice/IVoiceCommandManager.cs [hunk 4]
```diff
@@ -15,6 +15,15 @@ public interface IVoiceCommandManager
         /// <returns></returns>
         bool Register(string keyword, Action<string> @delegate);
 
+        /// <summary>
+        /// Registers a unprotected keyword for detection.
```
Comment (thegoldenmule) [inline Assets/Source/Input/Voice/IVoiceCommandManager.cs]: Nitpick: `a` should be `an`.

File: Assets/Source/Input/Voice/VoiceCommandManager.cs [hunk 5]
```diff
@@ -85,12 +96,17 @@ private struct KeywordConfig
...
+            return RegisterKeyword(keyword, callback, false, false);
         }
 
         public bool RegisterAdmin(string keyword, Action<string> callback)
         {
-            return RegisterKeyword(keyword, callback, true);
+            return RegisterKeyword(keyword, callback, true, false);
+        }
+
+        public bool RegisterUnprotected(string keyword, Action<string> callback)
```
Comment (thegoldenmule) [inline Assets/Source/Input/Voice/VoiceCommandManager.cs]: Missing comment on this function.

File: Assets/Source/Input/Voice/VoiceCommandManager.cs [hunk 6]
```diff
@@ -174,6 +192,28 @@ private void RebuildAdminGuard()
             }
         }
 
+        /// <summary>
+        /// Rebuilds the unprotected guard.
+        /// </summary>
+        private void RebuildUnprotectedGuard()
```
Comment (thegoldenmule) [inline Assets/Source/Input/Voice/VoiceCommandManager.cs]: I'm not sure I understand the point of this-- why is there a guard at all on unprotected words? I think we should just not require the guard on uprotected phrases rather than using this hack.

File: Assets/Source/Input/Voice/VoiceCommandManager.cs [hunk 7]
```diff
@@ -240,6 +289,14 @@ private void Recognizer_OnPhraseRecognized(PhraseRecognizedEventArgs args)
             KeywordConfig keywordConfig;
             if (_actions.TryGetValue(text, out keywordConfig))
             {
+
+                // check unprotected
+
```
Comment (thegoldenmule) [inline Assets/Source/Input/Voice/VoiceCommandManager.cs]: I think I see a pattern in the newlines-- these comments should not be flanked by newlines.

File: Assets/Source/Player/Scripting/Interface/VoiceJsInterface.cs [hunk 8]
```diff
@@ -90,6 +91,28 @@ public void registerAdmin(string command, IJsCallback callback)
             }
         }
 
+        /// <summary>
+        /// Registers an unprotected voice command to trigger a callback.
+        /// </summary>
+        /// <param name="engine"></param>
+        /// <param name="command"></param>
+        /// <param name="callback"></param>
+        public void registerProtected(string command, IJsCallback callback)
```
Comment (thegoldenmule) [inline Assets/Source/Player/Scripting/Interface/VoiceJsInterface.cs]: This should probably be `registerUnprotected`. However, I think the JS interface should actually _default_ to unprotected, now that we have it. Imagine a creator making an experience: why would they every want to have their players say "debug" + the command? I think this functions should be removed and the register method use the new `registerUnprotected`.

File: Assets/Source/Input/Voice/VoiceCommandManager.cs (hunk 7 above)
Comment (michaelcvoigt) [reply Assets/Source/Input/Voice/VoiceCommandManager.cs]: BTW, this is one of the files that doesn't connect to my VS solution. The text of the file is all grey... :/

Comment (thegoldenmule) [reply Assets/Source/Input/Voice/VoiceCommandManager.cs]: Ah, I see what you mean. If you look at the top of the file you can see this:

`#if NETFX_CORE || (!UNITY_EDITOR && UNITY_WSA)`

While in the editor, that will be false, meaning it compiles it out. Before Unity forced you to use IL2CPP, you could open in this in the generated VS solution when you built for WSA, but now I don't have a great solution for this, other than commenting out that line (which will show compile issues).

File: Assets/Source/Player/IUX/Content/ModelOutline.cs [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+                // Turn on alpha blending
+                _loadingLineMaterial.SetInt("_SrcBlend", (int)UnityEngine.Rendering.BlendMode.SrcAlpha);
+                _loadingLineMaterial.SetInt("_DstBlend", (int)UnityEngine.Rendering.BlendMode.OneMinusSrcAlpha);
+
+                // Turn backface culling off
+                _loadingLineMaterial.SetInt("_Cull", (int)UnityEngine.Rendering.CullMode.Off);
+
+                // Turn off depth writes
+                _loadingLineMaterial.SetInt("_ZWrite", 0);
+            }
```
Comment (jsundra) [inline Assets/Source/Player/IUX/Content/ModelOutline.cs]: `_loadingLineMaterial` is unused.

File: Assets/Source/Application/States/Play/Desktop/DesktopDesignController.cs [hunk 10]
```diff
@@ -279,14 +279,15 @@ private void SetupReferenceObject()
...
             if (null == outline)
             {
-                outline = _referenceCube.gameObject.AddComponent<ModelLoadingOutline>();
+                outline = _referenceCube.gameObject.AddComponent<ModelOutline>();
                 _referenceCube.gameObject.AddComponent<ReferenceObjectAxesRenderer>();
             }
 
-            outline.Init(bounds);
+            //Treat as selected to show this.
+            outline.Set(bounds,true);
```
[the remark this answers is not in the document]
Comment (thegoldenmule) [reply Assets/Source/Application/States/Play/Desktop/DesktopDesignController.cs]: Style: space after `//` in comment.

[the remark this answers is not in the document]
Comment (thegoldenmule) [reply Assets/Source/Application/States/Play/Desktop/DesktopDesignController.cs]: Also: I'm not loving the `Set` method here. The name and arguments are confusing-- is that a new thing?

Here's some good reading on the subject of booleans in method signatures:

https://ariya.io/2011/08/hall-of-api-shame-boolean-trap

TLDR: tread carefully. In this specific case I think we can do better.

File: Assets/Source/Application/Services/DebugService.cs [hunk 11]
```diff
@@ -142,18 +142,21 @@ public override void Start()
...
+            //_voice.Register("reset", Voice_OnReset);
+            //_voice.Register("update", Voice_OnUpdate);
+            //_voice.Register("experience", Voice_OnExperience);
+            //_voice.Register("network", Voice_OnNetwork);
+            //_voice.Register("anchors", Voice_OnAnchors);
+            //_voice.Register("performance", Voice_OnPerformance);
+            //_voice.Register("logging", Voice_OnLogging);
+            //_voice.Register("optimization", Voice_Optimization);
+            //_voice.Register("session", Voice_OnSession);
+            //_voice.RegisterAdmin("crash", _ => Log.Fatal(this, "Test crash."));
```
Comment (thegoldenmule) [inline Assets/Source/Application/Services/DebugService.cs]: Why are all of these commented out?

File: Assets/Prefabs/Vines/Common/Common.Confirmation.txt [hunk 12]
```diff
@@ -9,9 +9,9 @@
 	<Button
 		id='btn-confirm'
 		position=(-0.2,-0.3,0)
-		label='Confirm' />
+		label='Confirm' voiceActivator='CONFIRM' />
```
[the remark this answers is not in the document]
Comment (michaelcvoigt) [reply Assets/Prefabs/Vines/Common/Common.Confirmation.txt]: I have a dictionary of all the commands - 'VoiceKeywords_en_us..cs' 

Was following this model from that file :
   public const string EXIT = "exit";

EXIT is the Enum/Const idea ... 

But could go Pascal if we prefer that.

File: Assets/Source/Application/Services/DebugService.cs [hunk 13]
```diff
@@ -178,6 +181,8 @@ public override void Stop()
             _voice.Unregister("performance");
             _voice.Unregister("logging");
             _voice.Unregister("crash");
+            _voice.Unregister("edit");
```
[the remark this answers is not in the document]
Comment (michaelcvoigt) [reply Assets/Source/Application/Services/DebugService.cs]: Great point. I'm still getting my head around all the files that are involved with voice commands.

File: Assets/Source/Application/Services/DebugService.cs (hunk 11 above)
Comment (michaelcvoigt) [reply Assets/Source/Application/Services/DebugService.cs]: This is an artifact of WIP code. Definitely need to sort these out.

File: Assets/Source/Player/IUX/Content/ModelOutline.cs (hunk 9 above)
Comment (michaelcvoigt) [reply Assets/Source/Player/IUX/Content/ModelOutline.cs]: Great catch, I left this around to remind myself that there is a loading state that I current can't simulate. And I want to make sure it still is accounted for. 

I've disabled WiFi but can't get this state to occur. Any tricks on how to simulate it?

File: Assets/Source/Player/IUX/Activator/ActivatorPrimitive.cs [hunk 14]
```diff
@@ -436,6 +553,7 @@ private void InitializeProps()
             _propHighlighted = Schema.Get<bool>("highlighted");
             _propHighlightPriority = Schema.Get<int>("highlightPriority");
             _propInteractionEnabled = Schema.GetOwn("interactionEnabled", true);
+            _propVoiceActivator = Schema.Get<string>("voiceActivator");
```
[the remark this answers is not in the document]
Comment (michaelcvoigt) [reply Assets/Source/Player/IUX/Activator/ActivatorPrimitive.cs]: Changing that created VS to show an error with a potential fix.

File: Assets/Source/Player/IUX/Content/ModelOutline.cs (hunk 9 above)
Comment (jsundra) [reply Assets/Source/Player/IUX/Content/ModelOutline.cs]: You could try throwing in some artificial delay into the loading process. For instance there's a DelayedScriptCache that loads things as fast as it can, then spins a bit before actually pushing things into the app.

Also as a note, any assets downloaded are cached - so trying to throttle your internet connection or cause it to fail might not always work for testing if an asset is available locally.

File: Assets/Source/Input/Voice/VoiceCommandManager.cs [hunk 15]
```diff
@@ -212,41 +199,29 @@ private void DebugGuard_OnPhraseRecognized(PhraseRecognizedEventArgs args)
...
         private void Recognizer_OnPhraseRecognized(PhraseRecognizedEventArgs args)
         {
-            var text = args.text;
-
+            // Play a sound when command is recognized.
+            AudioClip heardCommand = _widgetConfig.HeardCommand;
+            AudioSource cameraSource = _widgetConfig.CamerAudioSource;
+
+            cameraSource.clip = heardCommand;
+            cameraSource.Play();
```
[the remark this answers is not in the document]
Comment (michaelcvoigt) [reply Assets/Source/Input/Voice/VoiceCommandManager.cs]: `        private void Recognizer_OnPhraseRecognized(PhraseRecognizedEventArgs args)
        {
            var text = args.text;

            KeywordConfig keywordConfig;
            if (_actions.TryGetValue(text, out keywordConfig))
            {
                // check unprotected
                if (keywordConfig.UnprotectedPhrase)
                { 
                    keywordConfig.Callback(text);
                    PlayCommandExecutedSound();
                }else
                {
                    // check voice lock
                    if (!_config.Debug.DisableVoiceLock)
                    {
                        if (DateTime.Now.Subtract(_debugGuardCalled).TotalSeconds > VOICE_LOCK_THRESHOLD_SECS)
                        {
                            return;
                        }
                    }
                    keywordConfig.Callback(text);
                    PlayCommandExecutedSound();
                }
            }
        }`

File: Assets/Editor/Tools/EditorWindow/Voice/VoiceCommandEditorWindow.cs [hunk 16]
```diff
@@ -48,42 +63,53 @@ private void OnGUI()
...
+                countY++;
 
-                    if (GUILayout.Button("Trigger"))
-                    {
-                        manager.Call(registration);
-                    }
+                if (NUM_ROWS <= countY)
+                {
+                    countY = 0;
+                    countX += BUTTON_WIDTH;
```
[the remark this answers is not in the document]
Comment (michaelcvoigt) [reply Assets/Editor/Tools/EditorWindow/Voice/VoiceCommandEditorWindow.cs]: Changed the window to dynamically scroll.

File: Assets/Source/Application/States/Play/Hmd/States/MainDesignState.cs [hunk 17]
```diff
@@ -329,17 +317,17 @@ private void Voice_OnSelect(string command)
 
             var elements = allElementDesignControllers.GetAllElementDesignControllers();
 
-            foreach (var element in elements)
-            {
-                if (element)
+            for (var i = 0; i < elements.Count; i++)
+            { 
+                if (elements[i])
```
Comment (thegoldenmule) [inline Assets/Source/Application/States/Play/Hmd/States/MainDesignState.cs]: Why is this necessary? This is a little bit of a red flag for me. Why is `GetAllElementDesignControllers` allowed to return bad controllers? This looks like it hides an underlying issue where we are not correctly handling design controllers.

File: Assets/Source/Application/States/Play/Hmd/Manager/ElementControllerGroup.cs [hunk 18]
```diff
@@ -73,14 +72,16 @@ private class ControllerBinding
         /// <returns>All of the Element Design Controllers.</returns>
         public List<ElementDesignController> GetAllElementDesignControllers()
```
Comment (thegoldenmule) [inline Assets/Source/Application/States/Play/Hmd/Manager/ElementControllerGroup.cs]: This name is too verbose. We know an `ElementControllerGroup` is a group of `ElementControllers`, so why not just call this method `All`?

File: Assets/Source/Application/States/Play/Hmd/States/MainDesignState.cs (hunk 17 above)
Comment (thegoldenmule) [reply Assets/Source/Application/States/Play/Hmd/States/MainDesignState.cs]: Also: it looks like you've added a method to get all design controllers just for this. Why not just make `FindClosest` a method on the `ControllerGroup`?

File: Assets/Source/Input/Voice/VoiceCommandManager.cs [hunk 19]
```diff
@@ -199,41 +199,48 @@ private void DebugGuard_OnPhraseRecognized(PhraseRecognizedEventArgs args)
...
         }
 
+        /// <summary>
+        /// Play the sound that signifies the command was understood and executed.
+        /// </summary>
+        private void PlayCommandExecutedSound()
+        {
+                    // Play a sound when command is recognized.
+                    AudioClip heardCommand = _widgetConfig.HeardCommand;
+                    AudioSource cameraSource = _widgetConfig.CameraAudioSource;
```
Comment (thegoldenmule) [inline Assets/Source/Input/Voice/VoiceCommandManager.cs]: vars

File: Assets/Source/Player/IUX/Widget/TextWidget.cs [hunk 20]
```diff
@@ -98,6 +99,9 @@ protected override void LoadInternalAfterChildren()
             _lineSpacingProp = Schema.Get<float>("lineSpacing");
             _lineSpacingProp.OnChanged += LineSpacing_OnChanged;
 
+            _fontColorProp = Schema.Get<string>("fontColor");
+            _fontColorProp.OnChanged += FontColor_OnChanged;
```
[the remark this answers is not in the document]
Comment (michaelcvoigt) [reply Assets/Source/Player/IUX/Widget/TextWidget.cs]: This passes the value down to the text primitive.

File: Assets/Source/Player/IUX/Widget/Select/SelectWidget.cs [hunk 21]
```diff
@@ -141,8 +172,12 @@ protected override void LoadInternalBeforeChildren()
                 _fontSizeProp = Schema.Get<int>("fontSize");
                 _fontSizeProp.OnChanged += FontSize_OnChanged;
 
+                _fontColorProp = Schema.Get<string>("fontColor");
+                _fontColorProp.OnChanged += FontColor_OnChanged;
```
[the remark this answers is not in the document]
Comment (michaelcvoigt) [reply Assets/Source/Player/IUX/Widget/Select/SelectWidget.cs]: Just tested and yes, this is what makes the text on the select widget have a color. I'm using it for the object browser, in this build the test is a slight blue color from the config...

File: Assets/Source/Player/IUX/Primitive/TextPrimitive.cs [hunk 22]
```diff
@@ -234,6 +240,28 @@ public float LineSpacing
...
+            set
+            {
+                Col4 fontColor4;
+                if (_colors.TryGetColor(value, out fontColor4))
+                {
+                    var fontColor = new Color(fontColor4.r, fontColor4.g, fontColor4.b, fontColor4.a);
+                    _renderer.Text.color = fontColor;
+                }
+
+                _propFontColor = Schema.Get<string>(value);
```
Comment (jsundra) [inline Assets/Source/Player/IUX/Primitive/TextPrimitive.cs]: This seems hacky. `_propFontColor` is being used instead of just a cached string, and is using Schema in a round about way.

If this prop followed the pattern of the other props, getting an upstream Schema prop, setting some formatting based on it, and observing changes - that type of code wouldn't need to be duplicated in things like `SelectWidget`.

With the text-effects branch into develop now, is all of this necessary still now that we can support `Col4`s for colors and not just the 12 baked in ones?

File: Assets/Source/Player/IUX/Primitive/TextPrimitive.cs [hunk 23]
```diff
@@ -278,15 +306,21 @@ public new Col4 Color
             ColorConfig colors,
             IMaterialManager materialManager)
             : base(
-                new GameObject("Text"),
+                new GameObject("text"),
```
Comment (jsundra) [inline Assets/Source/Player/IUX/Primitive/TextPrimitive.cs]: The rest of the codebase uses PascalCase for GameObject naming.

File: Assets/Source/Player/IUX/Widget/Select/SelectWidget.cs [hunk 24]
```diff
@@ -303,6 +338,36 @@ private void Left_OnActivated(ActivatorPrimitive activatorPrimitive)
...
+            string prev,
+            string next)
+        {
+            _text.FontColor = next;
+        }
+
+        /// <summary>
+        /// Called when the voice activator for the right button is changed.
+        /// </summary>
+        private void rightVoiceCommand_OnChanged(string command)
```
Comment (jsundra) [inline Assets/Source/Player/IUX/Widget/Select/SelectWidget.cs]: Method names start capitalized.

File: Assets/Source/Player/IUX/Widget/Select/SelectWidget.cs [hunk 25]
```diff
@@ -303,6 +338,36 @@ private void Left_OnActivated(ActivatorPrimitive activatorPrimitive)
...
+        {
+            _rightActivator.VoiceActivator = command;
+        }
+
+        /// <summary>
+        /// Called when the voice activator for the left button is changed.
+        /// </summary>
+        private void leftVoiceCommand_OnChanged(string command)
```
Comment (jsundra) [inline Assets/Source/Player/IUX/Widget/Select/SelectWidget.cs]: Method names start capitalized.

File: Assets/Source/Player/IUX/Primitive/TextPrimitive.cs (hunk 23 above)
Comment (michaelcvoigt) [reply Assets/Source/Player/IUX/Primitive/TextPrimitive.cs]: Changed and don't think this will cause an error. Searched the codebase for occurrences and didn't find any.

File: Assets/Source/Player/IUX/Primitive/TextPrimitive.cs (hunk 22 above)
Comment (michaelcvoigt) [reply Assets/Source/Player/IUX/Primitive/TextPrimitive.cs]: Can you suggest how to change this? @jsundra

Comment (jsundra) [reply Assets/Source/Player/IUX/Primitive/TextPrimitive.cs]: I was going to reference some other parts of `TextWidget` and `TextPrimitive`.... but those also are using Schema in weird ways. Soooooooo let's table this for later. 

MERGE IT. YOLO.
